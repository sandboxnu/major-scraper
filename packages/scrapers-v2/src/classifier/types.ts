export enum CatalogEntryType {
  Major = "Major",
  Minor = "Minor",
  Concentration = "Concentration",
  Unknown = "Unknown",
}

export type TypedCatalogEntry = { url: string; type: CatalogEntryType };
