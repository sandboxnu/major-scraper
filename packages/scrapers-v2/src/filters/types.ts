export enum CatalogEntryType {
  Major = "Major",
  Minor = "Minor",
  Concentration = "Concentration",
  Unknown = "Unknown",
}
