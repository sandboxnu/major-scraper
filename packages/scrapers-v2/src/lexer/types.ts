/**
 * Catalog Representations:
 *
 * { document { <section> ... } }
 * { section <desc> <comment> <row[]> }
 * { row { <textRow | courseRow | multiCourseRow> ...} }
 *    textRow:
 *      { areaHeader <comment> <hour> }
 *      { commentRow <comment> <hour> }
 *      { subHeader  <comment> <hour> }
 *
 *    courseRow:
 *      { courseRow <code> <hour> }
 *      { orCourseRow <code> <hour> }
 *
 *    multiCourseRow:
 *      { andCourseRow <hour> { <code, description> ... } }
 *
 *    rangeRow:
 *      { boundedRangeRow <hour> <subject> <startId> <endId> }
 *      { unboundedRangeRow <hour> <subj> }
 *      { lowerBoundedRangeRow <hour> <subject> <startId> }
 */

export type HDocument = {
  yearVersion: number;
  majorName: string;
  programRequiredHours: number;
  sections: HSection[];
};

export type HSection = {
  description: string;
  entries: HRow[];
};

export type HRow =
  | TextRow<HRowType.COMMENT>
  | TextRow<HRowType.HEADER>
  | TextRow<HRowType.SUBHEADER>
  | CourseRow<HRowType.OR_COURSE>
  | CourseRow<HRowType.PLAIN_COURSE>
  | MultiCourseRow<HRowType.AND_COURSE>
  | RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED>
  | WithExceptions<
      RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS>
    >
  | RangeBoundedRow<HRowType.RANGE_BOUNDED>
  | RangeUnboundedRow<HRowType.RANGE_UNBOUNDED>;

// the different outputs we have, by TYPE
export enum HRowType {
  HEADER = "HEADER",
  SUBHEADER = "SUBHEADER",
  COMMENT = "COMMENT",
  OR_COURSE = "OR_COURSE",
  AND_COURSE = "AND_COURSE",
  PLAIN_COURSE = "PLAIN_COURSE",

  RANGE_LOWER_BOUNDED = "RANGE_LOWER_BOUNDED",
  RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS = "RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS",

  RANGE_BOUNDED = "RANGE_BOUNDED",
  RANGE_UNBOUNDED = "RANGE_UNBOUNDED",
}

export interface TextRow<T> {
  type: T;
  description: string;
  hour: number;
}

export interface CourseRow<T> {
  type: T;
  description: string;
  hour: number;
  subject: string;
  classId: number;
}

export interface MultiCourseRow<T> {
  type: T;
  description: string;
  hour: number;
  // may contain duplicates
  courses: Array<{ subject: string; classId: number; description: string }>;
}

export interface RangeBoundedRow<T> {
  type: T;
  hour: number;
  subject: string;
  classIdStart: number;
  classIdEnd: number;
}

export interface RangeUnboundedRow<T> {
  type: T;
  hour: number;
  subjects: Array<string>;
}

export interface RangeLowerBoundedRow<T> {
  type: T;
  hour: number;
  subject: string;
  classIdStart: number;
}

export type WithExceptions<S> = S & {
  exceptions: Array<{
    subject: string;
    classId: number;
  }>;
};
