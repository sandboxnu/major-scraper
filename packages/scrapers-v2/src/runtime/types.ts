import { Result } from "@graduate/common";

export enum Phase {
  Urls = "Urls",
  Flatten = "Flatten",
  Classify = "Classify",
  Filter = "Filter",
  Tokenize = "Tokenize",
}

export type Pipeline<T> = {
  id: URL;
  trace: Phase[];
  result: Result<T, unknown[]>;
};
