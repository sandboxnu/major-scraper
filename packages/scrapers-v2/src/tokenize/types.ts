/**
 * An HTML document (catalog page) has a few identifiable features, along with a
 * bunch of sections.
 */
export type HDocument = {
  yearVersion: number;
  majorName: string;
  programRequiredHours: number;
  sections: HSection[];
};

/** An HTML section (of a document) has a description, and a list of rows that it contains. */
export type HSection = {
  description: string;
  entries: HRow[];
};

/**
 * An HTML row (abbreviated HRow) consists of four main different types of row:
 *
 * - TextRow: a row containing text. either an areaHeader, comment, or subHeader.
 * - CourseRow: a single course, that may have an "OR" annotation
 * - MultiCourseRow: multiple courses (2+). currently only AND courses appear as
 *   multiCourseRows
 * - RangeRow: either bounded, unbounded, or only bounded on the bottom (sometimes
 *   with exceptions)
 */
export type HRow =
  // text rows
  | TextRow<HRowType.COMMENT>
  | TextRow<HRowType.HEADER>
  | TextRow<HRowType.SUBHEADER>
  // course rows
  | CourseRow<HRowType.OR_COURSE>
  | CourseRow<HRowType.PLAIN_COURSE>
  // multi course rows
  | MultiCourseRow<HRowType.AND_COURSE>
  // range rows
  | RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED>
  | WithExceptions<
      RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS>
    >
  | RangeBoundedRow<HRowType.RANGE_BOUNDED>
  | WithExceptions<RangeBoundedRow<HRowType.RANGE_BOUNDED_WITH_EXCEPTIONS>>
  | RangeUnboundedRow<HRowType.RANGE_UNBOUNDED>;

// an enum to give a unique discriminator to each of the above cases
// the different outputs we have, by TYPE
export enum HRowType {
  HEADER = "HEADER",
  SUBHEADER = "SUBHEADER",
  COMMENT = "COMMENT",
  OR_COURSE = "OR_COURSE",
  AND_COURSE = "AND_COURSE",
  PLAIN_COURSE = "PLAIN_COURSE",

  RANGE_LOWER_BOUNDED = "RANGE_LOWER_BOUNDED",
  RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS = "RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS",

  RANGE_BOUNDED = "RANGE_BOUNDED",
  RANGE_BOUNDED_WITH_EXCEPTIONS = "RANGE_BOUNDED_WITH_EXCEPTIONS",

  RANGE_UNBOUNDED = "RANGE_UNBOUNDED",
}

export interface TextRow<T> {
  type: T;
  description: string;
  hour: number;
}

export interface CourseRow<T> {
  type: T;
  description: string;
  hour: number;
  subject: string;
  classId: number;
}

export interface MultiCourseRow<T> {
  type: T;
  description: string;
  hour: number;
  // may contain duplicates
  courses: Array<{ subject: string; classId: number; description: string }>;
}

export interface RangeBoundedRow<T> {
  type: T;
  hour: number;
  subject: string;
  classIdStart: number;
  classIdEnd: number;
}

export interface RangeUnboundedRow<T> {
  type: T;
  hour: number;
  subjects: Array<string>;
}

export interface RangeLowerBoundedRow<T> {
  type: T;
  hour: number;
  subject: string;
  classIdStart: number;
}

export type WithExceptions<S> = S & {
  exceptions: Array<{
    subject: string;
    classId: number;
  }>;
};
