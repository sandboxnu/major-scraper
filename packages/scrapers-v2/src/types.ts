/**
 * Catalog Representations:
 *
 * { catalog { <courseList> ... } }
 * { courseList <desc> <comment> <courseBody> } - table
 * { courseBody { <courseRow | textRow> ...} }
 *    textRow:
 *      { areaHeader <comment> <hour> }
 *      { commentRow <comment> <hour> }
 *      { subHeader  <comment> <hour> }
 *
 *    courseRow:
 *      { courseRow <codehol> <hour> }
 *      { orCourseRow <codehol> <hour> }
 *
 */

export type HTMLCatalog = {
  yearVersion: number;
  majorName: string;
  prgramRequiredHours: number;
  courseLists: HTMLCatalogCourseList[];
};

export type HTMLCatalogCourseList = {
  description: string;
  courseBody: HTMLCatalogCourseListBodyRow[];
};

export interface HTMLCatalogCourseListBodyRow {
  description: string;
  hour: number;
  courseTitle?: string;
  type: CourseListBodyRowType;
}

export enum CourseListBodyRowType {
  HEADER = "HEADER",
  SUBHEADER = "SUBHEADER",
  COMMENT = "COMMENT",
  OR_COURSE = "OR_COURSE",
  PLAIN_COURSE = "PLAIN_COURSE",
}
