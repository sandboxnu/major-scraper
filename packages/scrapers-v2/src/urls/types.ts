export enum College {
  ARTS_MEDIA_DESIGN = "arts-media-design",
  BUSINESS = "business",
  KHOURY = "computer-information-science",
  ENGINEERING = "engineering",
  HEALTH_SCIENCES = "health-sciences",
  SCIENCE = "science",
  SOCIAL_SCIENCES_HUMANITIES = "social-sciences-humanities",
}

type CollegeKeys = keyof typeof College;

export type AvailableMajors = {
  [key in CollegeKeys]: Array<string>;
};
