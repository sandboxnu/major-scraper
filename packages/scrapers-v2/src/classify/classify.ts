import { ensureLengthAtLeast, loadHTML, parseText } from "../utils";
import { CatalogEntryType, TypedCatalogEntry } from "./types";

export const addTypeToUrl = async (url: URL): Promise<TypedCatalogEntry> => {
  const type = getUrlType(await loadHTML(url.href));
  return { url, type };
};

const getUrlType = ($: CheerioStatic): CatalogEntryType => {
  const tabsContainer = $("#contentarea #tabs");

  if (tabsContainer.length === 0) {
    return CatalogEntryType.Unknown;
  } else if (tabsContainer.length === 1) {
    return getUrlTypeFromTabs($, tabsContainer.find("ul > li"));
  }
  throw new Error(
    `Expected 1 tab container, but found ${tabsContainer.length}.`
  );
};

const getUrlTypeFromTabs = ($: CheerioStatic, tabs: Cheerio) => {
  const [, middleTab] = ensureLengthAtLeast(2, tabs.toArray().map($));
  const middleTabText = parseText(middleTab);

  // most entries have 3 tabs, but some have 2 or rarely 4
  if (middleTabText === "Minor Requirements") {
    return CatalogEntryType.Minor;
  } else if (middleTabText === "Concentration Requirements") {
    return CatalogEntryType.Concentration;
  } else if (middleTabText === "Program Requirements") {
    return CatalogEntryType.Major;
  }

  throw new Error(
    `Middle tab text did not match one of the expected types: ${middleTabText}`
  );
};
