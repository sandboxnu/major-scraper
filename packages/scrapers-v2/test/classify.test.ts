import { addTypeToUrl } from "../src/classify/classify";
import {
  ACCEL_DEG_PROG_CAMD,
  ACCEL_DEG_PROG_COS,
  ACCEL_DEG_PROG_KHOURY,
  ACCOUNTING_MINOR,
  AEROSPACE_MINOR,
  ARCH_ENGLISH,
  ARCHITECTURE_MINOR,
  BSCS,
  BUSINESS,
  CHEMICAL_ENG,
  CS_GAME_DEV,
  CS_HISTORY,
  CS_MUSIC_WITH_TECH_MAJOR,
  DIGITAL_METHODS_HUMANITIES_MINOR,
  FINTECH_CONCENTRATION,
  GLOBAL_BUS_STRATEGY_MINOR,
  MATH_POLYSCI,
  MEDIA_SCREEN_STUDIES_HISTORY,
  WOMEN_GENDER_SEXUALITY_MINOR,
} from "./testUrls";
import { CatalogEntryType } from "../src/classify/types";

const MAJORS = [
  BSCS,
  CS_HISTORY,
  MEDIA_SCREEN_STUDIES_HISTORY,
  CS_MUSIC_WITH_TECH_MAJOR,
  BUSINESS,
  CHEMICAL_ENG,
  CS_GAME_DEV,

  // no tabs
  MATH_POLYSCI,
  ARCH_ENGLISH,

  // ending is in uppercase
  new URL(
    "https://catalog.northeastern.edu/undergraduate/science/behavioral-neuroscience/behavioral-neuroscience-philosophy-bs/"
  ),
];

const CONCENTRATIONS = [
  FINTECH_CONCENTRATION,

  // no tabs
  GLOBAL_BUS_STRATEGY_MINOR,

  // no comma in title
  new URL(
    "https://catalog.northeastern.edu/undergraduate/business/concentrations/accounting"
  ),
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/brand-management',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/business-analytics',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/corporate-innovation-venture',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/entrepreneurial-startups',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/family-business',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/fintech',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/finance',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/international-business',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/healthcare-management-consulting',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/management',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/management-information-systems',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/marketing',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/marketing-analytics',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/social-innovation-entrepreneurship',
  // 'https://catalog.northeastern.edu/undergraduate/business/concentrations/supply-chain-management'
];

const MINORS = [
  ACCOUNTING_MINOR,
  ARCHITECTURE_MINOR,
  AEROSPACE_MINOR,
  WOMEN_GENDER_SEXUALITY_MINOR,

  // no tabs
  DIGITAL_METHODS_HUMANITIES_MINOR,

  // has tabs, but with second tab text "program requirements" (not "minor requirements")
  new URL(
    "https://catalog.northeastern.edu/undergraduate/arts-media-design/communication-studies/sports-media-communication-minor"
  ),
  new URL(
    "https://catalog.northeastern.edu/undergraduate/arts-media-design/theatre/performing-arts-minor"
  ),
  new URL(
    "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/corporate-innovation-venture-minor"
  ),
  // "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/consulting-minor",
  // "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/entrepreneurial-startups-minor",
  // "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/family-business-minor",
  // "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/social-innovation-entrepreneurship-minor",
  // "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/strategy-minor",
  // "https://catalog.northeastern.edu/undergraduate/engineering/electrical-computer/robotics-minor",
  // "https://catalog.northeastern.edu/undergraduate/health-sciences/clinical-rehabilitation-sciences/human-movement-science-minor",
  // "https://catalog.northeastern.edu/undergraduate/social-sciences-humanities/cultures-societies-global-studies/german-minor",
];

const UNKNOWN = [
  // no tabs
  // no elem w id ending in "requirementstextcontainer"
  ACCEL_DEG_PROG_KHOURY,
  ACCEL_DEG_PROG_CAMD,
  ACCEL_DEG_PROG_COS,
  // ACCEL_DEG_PROG_BUSINESS,
  // ACCEL_DEG_PROG_ENG,
  // ACCEL_DEG_PROG_CSSH,
  // ACCEL_DEG_PROG_CSSH,
];

const testUrlsMatchType = (type: CatalogEntryType, urls: URL[]) => {
  describe(type, () => {
    for (const url of urls) {
      test(url.href, async () => {
        const { type } = await addTypeToUrl(url);
        expect(type).toBe(type);
      });
    }
  });
};

describe("classify", () => {
  testUrlsMatchType(CatalogEntryType.Unknown, UNKNOWN);
  testUrlsMatchType(CatalogEntryType.Major, MAJORS);
  testUrlsMatchType(CatalogEntryType.Minor, MINORS);
  testUrlsMatchType(CatalogEntryType.Concentration, CONCENTRATIONS);
});
