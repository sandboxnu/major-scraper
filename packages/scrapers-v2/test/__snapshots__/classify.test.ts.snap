// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters work https://catalog.northeastern.edu/undergraduate/arts-media-design/architecture/architectural-design-minor#programrequirementstext 1`] = `
Object {
  "type": "Minor",
  "url": "https://catalog.northeastern.edu/undergraduate/arts-media-design/architecture/architectural-design-minor#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/arts-media-design/communication-studies/media-screen-studies-history-ba#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/arts-media-design/communication-studies/media-screen-studies-history-ba#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/business/business-administration-bsba#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/business/business-administration-bsba#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/business/concentrations/fintech#programrequirementstext 1`] = `
Object {
  "type": "Concentration",
  "url": "https://catalog.northeastern.edu/undergraduate/business/concentrations/fintech#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/accounting-advisory-services-minor#programrequirementstext 1`] = `
Object {
  "type": "Minor",
  "url": "https://catalog.northeastern.edu/undergraduate/business/interdisciplinary-minors/accounting-advisory-services-minor#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/computer-information-science/accelerated-bachelor-graduate-degree-programs#programrequirementstext 1`] = `
Object {
  "type": "Unknown",
  "url": "https://catalog.northeastern.edu/undergraduate/computer-information-science/accelerated-bachelor-graduate-degree-programs#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-concentration-music-composition-technology-bs#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-concentration-music-composition-technology-bs#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-game-development-bs#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-game-development-bs#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-history-bs#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-information-science-combined-majors/computer-science-history-bs#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-science/bscs#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/computer-information-science/computer-science/bscs#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/engineering/chemical/chemical-engineering-bsche#programrequirementstext 1`] = `
Object {
  "type": "Major",
  "url": "https://catalog.northeastern.edu/undergraduate/engineering/chemical/chemical-engineering-bsche#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/engineering/mechanical-industrial/aerospace-minor#programrequirementstext 1`] = `
Object {
  "type": "Minor",
  "url": "https://catalog.northeastern.edu/undergraduate/engineering/mechanical-industrial/aerospace-minor#programrequirementstext",
}
`;

exports[`Filters work https://catalog.northeastern.edu/undergraduate/social-sciences-humanities/interdisciplinary/womens-gender-sexuality-studies-minor#programrequirementstext 1`] = `
Object {
  "type": "Minor",
  "url": "https://catalog.northeastern.edu/undergraduate/social-sciences-humanities/interdisciplinary/womens-gender-sexuality-studies-minor#programrequirementstext",
}
`;
